metricflow/inference/solver/weighted_tree.py

```diff
@@ -0,0 +1,163 @@
+from collections import defaultdict
+from typing import Callable, Dict, List, Optional
+
+from metricflow.dataflow.sql_column import SqlColumn
+from metricflow.inference.models import (
+    InferenceResult,
+    InferenceSignal,
+    InferenceSignalConfidence,
+    InferenceSignalNode,
+    InferenceSignalType,
+)
+from metricflow.inference.solver.base import InferenceSolver
+
+
+NodeWeighterFunction = Callable[[InferenceSignalConfidence], int]
+
+
+class WeightedTypeTreeInferenceSolver(InferenceSolver):
+    """Assigns weights to each type in the column type tree and attemptins to traverse it using a weight percentage threshold."""
+
+    @staticmethod
+    def default_weighter_function(confidence: InferenceSignalConfidence) -> int:
+        """The default weighter function.
+
+        It assigns weights 1, 2, 3 and 5 for LOW, MEDIUM, HIGH and FOR_SURE confidences, respectively. It then sums
+        the weights of all provided confidence scores.
+        """
+        confidence_weight_map = {
+            InferenceSignalConfidence.LOW: 1,
+            InferenceSignalConfidence.MEDIUM: 2,
+            InferenceSignalConfidence.HIGH: 4,
+            InferenceSignalConfidence.FOR_SURE: 8,
+        }
+
+        return confidence_weight_map[confidence]
+
+    def __init__(
+        self, weight_percent_threshold: float = 0.6, weighter_function: Optional[NodeWeighterFunction] = None
+    ) -> None:
+        """Initialize the solver.
+
+        weight_percent_threshold: a number between 0.5 and 1. If a node's weight corresponds to a percentage
+            above this threshold with respect to its siblings' total weight sum, the solver will progress deeper
+            into the type tree, entering that node. If not, it stops at the parent.
+        weighter_function: a function that returns a weight given a confidence score. It will be used
+            to assign integer weights to each node in the type tree based on its input signals.
+        """
+        assert weight_percent_threshold >= 0.5 and weight_percent_threshold <= 1
```

Comment:
This should solve the boundary issue below:
```suggestion
        weight_percent_threshold: a number in (0.5, 1]. If a node's weight corresponds to a percentage
            above this threshold with respect to its siblings' total weight sum, the solver will progress deeper
            into the type tree, entering that node. If not, it stops at the parent.
        weighter_function: a function that returns a weight given a confidence score. It will be used
            to assign integer weights to each node in the type tree based on its input signals.
        """
        assert weight_percent_threshold > 0.5 and weight_percent_threshold <= 1, f"weight_percent_threshold is {weight_percent_threshold}, but it must be > 0.5 and <= 1!"
```